README.md
# Exception 文档说明

主要描述 Exception 模块的结构, 以及使用说明

## 1. Exception 定义
### 1.1. Exception class

通过继承 Node.js Error 来实现 checked exceptions 的定义. 以下为 Exception 字段构成描述.

#### Properties

* `code` - 自定义错误代码
* `message` - 错误信息
* `stack` - 错误堆栈信息

#### Methods

- `fetchPropertyStacks() Beta版接口`
	* 该接口是返回错误代码 `code`. 类型为数组, 它将返回所有异常链上的错误代码信息. 目前, 处于测试阶段, 各个业务层不需调用该方法.
	* 该接口返回异常的 `message` 信息. 类型为数组, 它将返回异常链上的错误信息. 目前, 处于测试阶段, 各个业务层不需调用该方法.
- `getCause()` 获取最终原因对象.
	
### 1.2 Http Response Exceptions

用于 Application 层定义需要返回的 Http Exception 类型. 对 Web Application 根据业务场景返回 Http Response 信息进行了封装. 应用层需自定义和业务相关的异常信息来继承以下对应的异常类型.
- `BadRequestException` - 400
- `UnauthorizedException` - 401
- `PaymentRequiredException` - 402
- `ForbiddenException` - 403
- `NotFoundException` - 404
- `MethodNotAllowedException` - 405 
- `NotAcceptableException` - 406
- `ConflictException` - 409
- `GoneException` - 410
- `PreconditionFailedException` - 412
- `RequestEntityTooLargeException` - 413
- `InternalServerErrorException` - 500
- `NotImplementedException` - 501
- `ServiceUnavailableException` - 503
目前, 提供了以上几种 Exceptions. 主要参考文章: [Exceptions and HTTP Status Codes](https://cloud.google.com/endpoints/docs/frameworks/java/exceptions)
	
## 2. 异常处理框架分析(Blog)

异常处理也具备框架性结构, 根据领域的架构分解, 我认为异常的处理也是分而治之的思想来处理. 所以在异常的处理上作出了一下的架构分析.

### 2.1. 异常处理的原则

异常信息必然是被**内部调用者**自内而外的传递给**外部调用者**. 所以, 必须明确 **异常 Payload** 所必须承载的异常信息.

#### 分析1: 用户请求处理

> Who cares what kind of outputting things?(谁关心什么样的输出).  -- Benjamin Woo

在实际的架构中, 我们需要处理用户从远端发起的请求(**Http Request**). 并且根据业务逻辑对用户的输入作出相应的结果响应.

#### 分析2: 业务场景驱动下的异常信息

> 异常设计的初衷一定是不会脱离实际的业务流程. -- Benjamin Woo

我们有一个已知必然存在的一个关键设计要素 ---- **业务流程**. 任何明确的业务流程里, 必然存在异常业务处理流程分支. 如果有以上一个设计要素存在, 那么就能够非常清晰的将业务异常处理流程转化为代码设计. 前提是我们已经明确了我们的需求, 明确了我们的业务流程.

#### 分析3: 软件架构中业务异常处理的拆分粒度

在软件代码设计环节, 我们需要将大量的业务流程转化成软件业务代码. 但这里是有一个在开发环节的诉求, 就是拥有一个合理的软件代码结构. 以便于能够很好的规划和适应日新月异的业务需求. 所以, 这里提出设计的关键要素是如何将业务流程转化到我们的软件架构中去.

##### 业务异常分支也是业务流程的一部分(异常情况下的应用场景)

业务异常分支也是业务流程的一部分, 表达了用户应用场景的一部分. 实际上我们不能够脱离需求谈代码设计.
实际上我们回头看看我们之前走过的, 确实是掉进了一个误区. 我们一直在极力去设计一个合理的 Exception 处理结构. 但始终没有找到合适的答案并总是在凑合着使用.


##### Exception 在软件架构中的实施
分析我们现有的软件架构 -- 领域模型. 
从现实出发, 分析核心的代码模块.

`Port.Adapter` - 我们提供 Swagger APIs, 提供对外的沟通渠道. 
`Application` - 提供应用业务流程, 根据应用业务流程提供异常流程分支处理.
`Domain` - 提供领域模型功能, 根据领域模型提供的功能提供异常流程分支处理.

#### 具体的案例分析




CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.1"></a>
## 0.1.1 (2018-07-18)



<a name="0.1.0"></a>
# 0.1.0 (2018-07-18)


lib/Exception.js
const _ = require('lodash');

const SYSTEM_ERROR_CODE = null;

class Exception extends Error {
  /**
   * @constructor
   * @param {Exception} cause - 引起异常的原因
   */
  constructor(cause) {
    super();
    /**
     * 将自定义 Exception 名称设定为 Exception name
     */
    this.name = this.constructor.name;
    /**
     * 设定默认的 Exception code 为 -1 作为默认系统错误, 该值会被自定义的 Exception code 覆盖.
     */
    this.code = SYSTEM_ERROR_CODE;
    /**
     * 设定默认的 Exception message, 该值会被自定义的 Exception message 覆盖.
     */
    this.message = 'Exception error';
    Error.captureStackTrace(this, Exception);
    /**
     * 目前, 异常系统设计涉及了"分层异常记录"的问题. 可以将分层设计的异常抛给上层, 由上层捕捉下层抛出的错误.
     */
    this.cause = cause;
    /**
     * 设定该 Exception 定义为 Checked Exception.
     */
    this.setCheckedFlag(cause);
  }
  /**
   * @description 设置 Checked Exception flag 值
   * @param {Exception} cause
   */
  setCheckedFlag(cause) {
    this.isChecked = true;
    /**
     * NOTE: 该代码为评估版代码, 根据具体应用情况进行更新.
     * 如果 cause 是一个 unchecked Exception, 则我们需要将 Exception checkedException 设定为 false 以便最终按照
     * Error 来处理. 并且记录 Cause 记录的 Error 信息.
     * 这里从设计上理解为: 在 Exception chains 中, 如果 cause 出现是 Error 的情况, 意指该异常链中出现了未能捕捉的错误.
     * 所以, 会影响整个链, 以致于最后抛出的错误被记录为 unchecked Exception.
     */
    if (cause) {
      if (!cause.isChecked || cause.isChecked === undefined) {
        this.isChecked = false;
      }
    }
  }
  set name(name) {
    this.__name = name;
  }
  get name() {
    return this.__name;
  }
  set code(code) {
    this.__code = code;
  }
  get code() {
    return this.__code;
  }
  set message(message) {
    this.__message = message;
  }
  get message() {
    return this.__message;
  }
  set cause(exception) {
    this.__cause = exception || null;
  }
  get cause() {
    return this.__cause;
  }
  set isChecked(val) {
    this.__isChecked = val;
  }
  get isChecked() {
    return this.__isChecked;
  }
  /**
   * @description 从 Exception chains 中获取对应属性的 stack 信息.
   * @param {string} property - Exception 对象属性.
   * @return {Array<string>}
   */
  fetchPropertyStacks(property) {
    const exception = this;
    return Exception.fetchStacks(exception, property, []);
  }
  /**
   * @description 获取 last cause exception.
   * @return {Exception}
   */
  getCause() {
    const exception = this;
    return Exception.getCauseException(exception);
  }

  /**
   * @description 获取 last cause exception.
   * @param {Exception} exception - 异常对象
   * @return {Exception}
   */
  static getCauseException(exception) {
    if (_.isEmpty(exception.cause)) {
      return exception;
    }
    return Exception.getCauseException(exception.cause);
  }

  /**
   * @description 从 Exception chains 中获取对应属性的 stack 信息.
   * @param {Exception} exception - 异常对象
   * @param {string} property - 异常对象中的属性
   * @param {Array} stacks - 用于存放属性堆栈信息数组.
   * @return {Array<*>}
   */
  static fetchStacks(exception, property, stacks) {
    if (_.isEmpty(exception)) {
      return stacks;
    }
    stacks.push(exception[property]);
    if (!_.isEmpty(exception.cause)) {
      this.fetchStacks(exception.cause, property, stacks);
    }
    return stacks;
  }

  /**
   * @description 判断该异常为 unchecked Exception
   * NOTE: 说明 isChecked 会出现 true, false, undefined(非 Exception Error,
   * 也就是非继承自 rms-exception 基类的 Error, 不存在 isChecked 该属性)三种状态.
   * @param exception
   */
  static isError(exception) {
    return !exception.isChecked;
  }
}

module.exports = Exception;


example/example.js
const { Exception } = require('../index');

class UserSigningInException extends Exception {
  constructor(cause = null) {
    super(cause);
    this.code = 'ACCOUNT_SIGN_IN';
    this.message = 'User sign in error';
  }
}

class InvalidUserPasswordException extends Exception {
  constructor(cause) {
    super(cause);
    this.code = 'INVALID_PASSWORD';
    this.message = 'Invalid user password';
  }
}

function validateUserPassword() {
  const a = undefined;
  a.b = '';
  throw new InvalidUserPasswordException();
}

function userSignIn() {
  try {
    validateUserPassword();
  } catch (exc) {
    throw new UserSigningInException(exc);
  }
}

function processUserSignInService() {
  try {
    return userSignIn();
  } catch (exc) {
    console.log(exc);
    console.log(exc.getCause());
    console.log(exc.fetchPropertyStacks('message'));
  }
  return 'Nothing';
}

processUserSignInService();


lib/HttpResponseExceptions.js
/* eslint-disable no-unused-vars */
const Exception = require('./Exception');
const HttpStatus = require('./HttpStatus');

/**
 * HTTP 4xx
 */
class BadRequestException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.BAD_REQUEST];
    this.httpStatus = HttpStatus.BAD_REQUEST;
    this.code = HttpStatus.BAD_REQUEST;
  }
}
class UnauthorizedException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.UNAUTHORIZED];
    this.httpStatus = HttpStatus.UNAUTHORIZED;
    this.code = HttpStatus.UNAUTHORIZED;
  }
}
class PaymentRequiredException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.PAYMENT_REQUIRED];
    this.httpStatus = HttpStatus.PAYMENT_REQUIRED;
    this.code = HttpStatus.PAYMENT_REQUIRED;
  }
}
class ForbiddenException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.FORBIDDEN];
    this.httpStatus = HttpStatus.FORBIDDEN;
    this.code = HttpStatus.FORBIDDEN;
  }
}
class NotFoundException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.NOT_FOUND];
    this.httpStatus = HttpStatus.NOT_FOUND;
    this.code = HttpStatus.NOT_FOUND;
  }
}
class MethodNotAllowedException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.METHOD_NOT_ALLOWED];
    this.httpStatus = HttpStatus.METHOD_NOT_ALLOWED;
    this.code = HttpStatus.METHOD_NOT_ALLOWED;
  }
}

class NotAcceptableException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.NOT_ACCEPTABLE];
    this.httpStatus = HttpStatus.NOT_ACCEPTABLE;
    this.code = HttpStatus.NOT_ACCEPTABLE;
  }
}
class ProxyAuthenticationRequiredException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.PROXY_AUTHENTICATION_REQUIRED];
    this.httpStatus = HttpStatus.PROXY_AUTHENTICATION_REQUIRED;
    this.code = HttpStatus.PROXY_AUTHENTICATION_REQUIRED;
  }
}
class RequestTimeoutException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.REQUEST_TIMEOUT];
    this.httpStatus = HttpStatus.REQUEST_TIMEOUT;
    this.code = HttpStatus.REQUEST_TIMEOUT;
  }
}
class ConflictException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.CONFLICT];
    this.httpStatus = HttpStatus.CONFLICT;
    this.code = HttpStatus.CONFLICT;
  }
}
class GoneException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.GONE];
    this.httpStatus = HttpStatus.GONE;
    this.code = HttpStatus.GONE;
  }
}
class LengthRequiredException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.LENGTH_REQUIRED];
    this.httpStatus = HttpStatus.LENGTH_REQUIRED;
    this.code = HttpStatus.LENGTH_REQUIRED;
  }
}
class PreconditionFailedException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.PRECONDITION_FAILED];
    this.httpStatus = HttpStatus.PRECONDITION_FAILED;
    this.code = HttpStatus.PRECONDITION_FAILED;
  }
}
class RequestEntityTooLargeException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.REQUEST_ENTITY_TOO_LARGE];
    this.httpStatus = HttpStatus.REQUEST_ENTITY_TOO_LARGE;
    this.code = HttpStatus.REQUEST_ENTITY_TOO_LARGE;
  }
}
class RequestUriTooLongException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.REQUEST_URI_TOO_LONG];
    this.httpStatus = HttpStatus.REQUEST_URI_TOO_LONG;
    this.code = HttpStatus.REQUEST_URI_TOO_LONG;
  }
}
class UnsupportedMediaTypeException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.UNSUPPORTED_MEDIA_TYPE];
    this.httpStatus = HttpStatus.UNSUPPORTED_MEDIA_TYPE;
    this.code = HttpStatus.UNSUPPORTED_MEDIA_TYPE;
  }
}
class RequestedRangeNotSatisfiableException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.REQUESTED_RANGE_NOT_SATISFIABLE];
    this.httpStatus = HttpStatus.REQUESTED_RANGE_NOT_SATISFIABLE;
    this.code = HttpStatus.REQUESTED_RANGE_NOT_SATISFIABLE;
  }
}
class ExceptionFailedException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.EXPECTATION_FAILED];
    this.httpStatus = HttpStatus.EXPECTATION_FAILED;
    this.code = HttpStatus.EXPECTATION_FAILED;
  }
}
class ImATeapotException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.IM_A_TEAPOT];
    this.httpStatus = HttpStatus.IM_A_TEAPOT;
    this.code = HttpStatus.IM_A_TEAPOT;
  }
}
class MisdirectedRequestException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.MISDIRECTED_REQUEST];
    this.httpStatus = HttpStatus.MISDIRECTED_REQUEST;
    this.code = HttpStatus.MISDIRECTED_REQUEST;
  }
}
class UnprocessableEntityException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.UNPROCESSABLE_ENTITY];
    this.httpStatus = HttpStatus.UNPROCESSABLE_ENTITY;
    this.code = HttpStatus.UNPROCESSABLE_ENTITY;
  }
}
class UpgradeRequiredException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.UPGRADE_REQUIRED];
    this.httpStatus = HttpStatus.UPGRADE_REQUIRED;
    this.code = HttpStatus.UPGRADE_REQUIRED;
  }
}
class PreconditionRequiredException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.PRECONDITION_REQUIRED];
    this.httpStatus = HttpStatus.PRECONDITION_REQUIRED;
    this.code = HttpStatus.PRECONDITION_REQUIRED;
  }
}
class LockedException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.LOCKED];
    this.httpStatus = HttpStatus.LOCKED;
    this.code = HttpStatus.LOCKED;
  }
}
class FailedDependencyException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.FAILED_DEPENDENCY];
    this.httpStatus = HttpStatus.FAILED_DEPENDENCY;
    this.code = HttpStatus.FAILED_DEPENDENCY;
  }
}
class TooManyRequestsException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.TOO_MANY_REQUESTS];
    this.httpStatus = HttpStatus.TOO_MANY_REQUESTS;
    this.code = HttpStatus.TOO_MANY_REQUESTS;
  }
}
class RequestHeaderFieldsTooLargeException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.REQUEST_HEADER_FIELDS_TOO_LARGE];
    this.httpStatus = HttpStatus.REQUEST_HEADER_FIELDS_TOO_LARGE;
    this.code = HttpStatus.REQUEST_HEADER_FIELDS_TOO_LARGE;
  }
}
class UnvailableForLegalReasonsException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.UNAVAILABLE_FOR_LEGAL_REASONS];
    this.httpStatus = HttpStatus.UNAVAILABLE_FOR_LEGAL_REASONS;
    this.code = HttpStatus.UNAVAILABLE_FOR_LEGAL_REASONS;
  }
}

/**
 * HTTP 5xx
 */

class InternalServerErrorException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.INTERNAL_SERVER_ERROR];
    this.httpStatus = HttpStatus.INTERNAL_SERVER_ERROR;
    this.code = HttpStatus.INTERNAL_SERVER_ERROR;
  }
}
class NotImplementedException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.NOT_IMPLEMENTED];
    this.httpStatus = HttpStatus.NOT_IMPLEMENTED;
    this.code = HttpStatus.NOT_IMPLEMENTED;
  }
}
class ServiceUnavailableException extends Exception {
  constructor(...args) {
    super(...args);
    this.message = HttpStatus[HttpStatus.SERVICE_UNAVAILABLE];
    this.httpStatus = HttpStatus.SERVICE_UNAVAILABLE;
    this.code = HttpStatus.SERVICE_UNAVAILABLE;
  }
}

module.exports = {
  BadRequestException,
  UnauthorizedException,
  PaymentRequiredException,
  ForbiddenException,
  NotFoundException,
  MethodNotAllowedException,
  NotAcceptableException,
  ConflictException,
  GoneException,
  PreconditionFailedException,
  RequestEntityTooLargeException,
  InternalServerErrorException,
  NotImplementedException,
  ServiceUnavailableException,
};


lib/HttpStatus.js
const HttpStatus = {
  100: 'Continue',
  101: 'Switching Protocols',
  200: 'OK',
  201: 'Created',
  202: 'Accepted',
  203: 'Non-Authoritative Information',
  204: 'No Content',
  205: 'Reset Content',
  206: 'Partial Content',
  207: 'Multi Status',
  208: 'Already Reported',
  226: 'IM Used',
  300: 'Multiple Choices',
  301: 'Moved Permanently',
  302: 'Found',
  303: 'See Other',
  304: 'Not Modified',
  305: 'Use Proxy',
  306: 'Switch Proxy',
  307: 'Temporary Redirect',
  308: 'Permanent Redirect',
  400: 'Bad Request',
  401: 'Unauthorized',
  402: 'Payment Required',
  403: 'Forbidden',
  404: 'Not Found',
  405: 'Method Not Allowed',
  406: 'Not Acceptable',
  407: 'Proxy Authentication Required',
  408: 'Request Time-out',
  409: 'Conflict',
  410: 'Gone',
  411: 'Length Required',
  412: 'Precondition Failed',
  413: 'Request Entity Too Large',
  414: 'Request-URI Too Large',
  415: 'Unsupported Media Type',
  416: 'Requested Range not Satisfiable',
  417: 'Expectation Failed',
  418: 'I\'m a teapot',
  421: 'Misdirected Request',
  422: 'Unprocessable Entity',
  423: 'Locked',
  424: 'Failed Dependency',
  426: 'Upgrade Required',
  428: 'Precondition Required',
  429: 'Too Many Requests',
  431: 'Request Header Fields Too Large',
  451: 'Unavailable For Legal Reasons',
  500: 'Internal Server Error',
  501: 'Not Implemented',
  502: 'Bad Gateway',
  503: 'Service Unavailable',
  504: 'Gateway Time-out',
  505: 'HTTP Version not Supported',
  506: 'Variant Also Negotiates',
  507: 'Insufficient Storage',
  508: 'Loop Detected',
  510: 'Not Extended',
  511: 'Network Authentication Required',
  CONTINUE: 100,
  SWITCHING_PROTOCOLS: 101,
  OK: 200,
  CREATED: 201,
  ACCEPTED: 202,
  NON_AUTHORITATIVE_INFORMATION: 203,
  NO_CONTENT: 204,
  RESET_CONTENT: 205,
  PARTIAL_CONTENT: 206,
  MULTI_STATUS: 207,
  ALREADY_REPORTED: 208,
  IM_USED: 226,
  MULTIPLE_CHOICES: 300,
  MOVED_PERMANENTLY: 301,
  FOUND: 302,
  SEE_OTHER: 303,
  NOT_MODIFIED: 304,
  USE_PROXY: 305,
  SWITCH_PROXY: 306,
  TEMPORARY_REDIRECT: 307,
  PERMANENT_REDIRECT: 308,
  BAD_REQUEST: 400,
  UNAUTHORIZED: 401,
  PAYMENT_REQUIRED: 402,
  FORBIDDEN: 403,
  NOT_FOUND: 404,
  METHOD_NOT_ALLOWED: 405,
  NOT_ACCEPTABLE: 406,
  PROXY_AUTHENTICATION_REQUIRED: 407,
  REQUEST_TIMEOUT: 408,
  CONFLICT: 409,
  GONE: 410,
  LENGTH_REQUIRED: 411,
  PRECONDITION_FAILED: 412,
  REQUEST_ENTITY_TOO_LARGE: 413,
  REQUEST_URI_TOO_LONG: 414,
  UNSUPPORTED_MEDIA_TYPE: 415,
  REQUESTED_RANGE_NOT_SATISFIABLE: 416,
  EXPECTATION_FAILED: 417,
  IM_A_TEAPOT: 418,
  MISDIRECTED_REQUEST: 421,
  UNPROCESSABLE_ENTITY: 422,
  UPGRADE_REQUIRED: 426,
  PRECONDITION_REQUIRED: 428,
  LOCKED: 423,
  FAILED_DEPENDENCY: 424,
  TOO_MANY_REQUESTS: 429,
  REQUEST_HEADER_FIELDS_TOO_LARGE: 431,
  UNAVAILABLE_FOR_LEGAL_REASONS: 451,
  INTERNAL_SERVER_ERROR: 500,
  NOT_IMPLEMENTED: 501,
  BAD_GATEWAY: 502,
  SERVICE_UNAVAILABLE: 503,
  GATEWAY_TIMEOUT: 504,
  HTTP_VERSION_NOT_SUPPORTED: 505,
  VARIANT_ALSO_NEGOTIATES: 506,
  INSUFFICIENT_STORAGE: 507,
  LOOP_DETECTED: 508,
  NOT_EXTENDED: 510,
  NETWORK_AUTHENTICATION_REQUIRED: 511,
};

module.exports = HttpStatus;
